app/api/odm/ModelWithPermissions.ts

```diff
@@ -0,0 +1,159 @@
+    const userIds = getUserPermissionIds(user);
+    const levelCond = level === AccessLevels.WRITE ? { level: AccessLevels.WRITE } : {};
+    permissionCond = {
+      $or: [
+        { permissions: { $elemMatch: { _id: { $in: userIds }, ...levelCond } } },
+        { published: true },
+      ],
+    };
+  }
+  return permissionCond;
+};
+
+const appendPermissionQuery = <T>(
+  query: PermissionsUwaziFilterQuery<T>,
+  level: AccessLevels,
+  user: UserSchema | undefined
+) => {
+  let permissionCond;
+  if (user) {
+    permissionCond = addPermissionsCondition(user, level);
+  } else if (level === AccessLevels.READ) {
+    permissionCond = { published: true };
+  } else {
+    permissionCond = { _id: null };
+  }
+  return { ...query, ...permissionCond };
+};
+
+function checkPermissionAccess<T>(
+  elem: T & { permissions?: PermissionSchema[] },
+  userIds: ObjectIdSchema[],
+  level: AccessLevels = AccessLevels.WRITE
+) {
+  const hasAccessLevel = (p: PermissionSchema) =>
+    level === AccessLevels.WRITE
+      ? p.level === AccessLevels.WRITE && userIds.includes(p._id)
+      : userIds.includes(p._id);
+
+  const hasAccess = elem.permissions && elem.permissions.find(p => hasAccessLevel(p)) !== undefined;
+  if (!hasAccess) {
+    return { ...elem, permissions: undefined };
+  }
+  return elem;
+}
+
+const filterPermissionsData = <T>(data: T[], user: UserSchema | undefined) => {
+  let filteredData = data;
+  if (user && !['admin', 'editor'].includes(user.role)) {
+    if (Array.isArray(data) && data.length > 0) {
+      const userIds = getUserPermissionIds(user);
+      filteredData = data.map(elem => checkPermissionAccess(elem, userIds));
+    }
+  }
+  return filteredData;
+};
+
+const controlPermissionsData = <T>(data: T & { published?: boolean }) => {
+  const user = permissionsContext.getUserInContext();
+  let controlledData = null;
+  if (user) {
+    if (['admin', 'editor'].includes(user.role)) {
+      controlledData = data;
+    } else {
+      const doc = checkPermissionAccess(data, getUserPermissionIds(user), AccessLevels.READ);
+      if (doc.permissions) {
+        controlledData = { ...data, permissions: undefined };
+      }
+    }
+  } else if (data.published) {
+    controlledData = { ...data, permissions: undefined };
+  }
+  return controlledData;
+};
+
+export class ModelWithPermissions<T> extends OdmModel<T> {
+  async save(data: DataType<T>) {
+    const user = permissionsContext.getUserInContext();
+    return data._id
+      ? super.save(data, appendPermissionQuery({ _id: data._id }, AccessLevels.WRITE, user))
+      : super.save(appendPermissionData(data, user));
+  }
+
+  get(query: UwaziFilterQuery<T> = {}, select: any = '', options: {} = {}) {
+    const user = permissionsContext.getUserInContext();
+    const results = super.get(
+      appendPermissionQuery(query, AccessLevels.READ, user),
+      select,
+      options
+    );
+    return results.map(data => filterPermissionsData(data, user));
```

Comment:
Yes, your approach works, we just wanted to avoid a second query. Anyways, if you run a second query you still need to iterate over the set to append the permissions. In the other hand, if you are thinking about requesting the permissions WHEN needed, like, in a second end point, then it'll change the whole architecture.

Reply:
we agreed to make the permissions not being returned by default, but there should be the ability to ask for them, this will trigger the filterPermissions post process.
the way to ask for them is using the same select object from mongo